compiler/rustc_ast/src/ast.rs

```diff
@@ -550,6 +550,8 @@ pub struct Pat {
     pub id: NodeId,
     pub kind: PatKind,
     pub span: Span,
+    /// Whether or not this pattern starts with a leading `|`
+    pub leading_vert: bool,
```

Comment:
is `vert` terminology used elsewhere? Looking at the two `TokenKind`s (one in `ast`, one in `rustc_lexer`), the name for the thing is `Or`, so maybe stick with `leading_or`?

OTOH, I don't really enjoy `and` / `or` names for `&` and `|` symbols, so we might want to rename those (obviously, in a separate PR). In rust-analyzer, those are called `amp` and `pipe`, but `vert` or `vbar` also would work for `|` I think (although i do find `vert` specifically confusing a bit).

Reply:
The `vert` terminology is used in `parse_top_pat`, which is why I picked it.